core/drivers/pl022_spi.c

```diff
@@ -0,0 +1,505 @@
+#define SSPICR_RORIC	SHIFT_U32(1, 0)
+
+#define SSPDMACR_TXDMAE	SHIFT_U32(1, 1)
+#define SSPDMACR_RXDMAE	SHIFT_U32(1, 0)
+
+#define SSPPeriphID0_PartNumber0	SHIFT_U32(0xFF, 0) /* 0x22 */
+#define SSPPeriphID1_Designer0		SHIFT_U32(0xF, 4) /* 0x1 */
+#define SSPPeriphID1_PartNumber1	SHIFT_U32(0xF, 0) /* 0x0 */
+#define SSPPeriphID2_Revision		SHIFT_U32(0xF, 4)
+#define SSPPeriphID2_Designer1		SHIFT_U32(0xF, 0) /* 0x4 */
+#define SSPPeriphID3_Configuration	SHIFT_U32(0xFF, 0) /* 0x00 */
+
+#define SSPPCellID_0	SHIFT_U32(0xFF, 0) /* 0x0D */
+#define SSPPCellID_1	SHIFT_U32(0xFF, 0) /* 0xF0 */
+#define SSPPPCellID_2	SHIFT_U32(0xFF, 0) /* 0x05 */
+#define SSPPPCellID_3	SHIFT_U32(0xFF, 0) /* 0xB1 */
+
+#define MASK_32 0xFFFFFFFF
+#define MASK_28 0xFFFFFFF
+#define MASK_24 0xFFFFFF
+#define MASK_20 0xFFFFF
+#define MASK_16 0xFFFF
+#define MASK_12 0xFFF
+#define MASK_8 0xFF
+#define MASK_4 0xF
+/* spi register masks */
+
+#define SSP_CPSDVR_MAX		254
+#define SSP_CPSDVR_MIN		2
+#define SSP_SCR_MAX		255
+#define SSP_SCR_MIN		0
+#define SSP_DATASIZE_MAX	16
+
+enum pl022_data_size {
+	PL022_DATA_SIZE4 = 0x3,
+	PL022_DATA_SIZE5,
+	PL022_DATA_SIZE6,
+	PL022_DATA_SIZE7,
+	PL022_DATA_SIZE8 = SSPCR0_DSS_8BIT,
+	PL022_DATA_SIZE9,
+	PL022_DATA_SIZE10,
+	PL022_DATA_SIZE11,
+	PL022_DATA_SIZE12,
+	PL022_DATA_SIZE13,
+	PL022_DATA_SIZE14,
+	PL022_DATA_SIZE15,
+	PL022_DATA_SIZE16 = SSPCR0_DSS_16BIT
+};
+
+enum pl022_spi_mode {
+	PL022_SPI_MODE0 = SSPCR0_SPO0 | SSPCR0_SPH0, /* 0x00 */
+	PL022_SPI_MODE1 = SSPCR0_SPO0 | SSPCR0_SPH1, /* 0x80 */
+	PL022_SPI_MODE2 = SSPCR0_SPO1 | SSPCR0_SPH0, /* 0x40 */
+	PL022_SPI_MODE3 = SSPCR0_SPO1 | SSPCR0_SPH1  /* 0xC0 */
+};
+
+static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat,
+	uint8_t *rdat, size_t num_txpkts, size_t *num_rxpkts)
+{
+	size_t i = 0, j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	while (i < num_txpkts) {
+		if (read8(pd->base + SSPSR) & SSPSR_TNF)
+			/* tx 1 packet */
+			write8(wdat[i++], pd->base + SSPDR);
+	}
+
+	do {
+		while ((read8(pd->base + SSPSR) & SSPSR_RNE)
+			&& (j < *num_rxpkts))
+			/* rx 1 packet */
+			rdat[j++] = read8(pd->base + SSPDR);
+	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
+
+	*num_rxpkts = j;
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat,
+	uint16_t *rdat, size_t num_txpkts, size_t *num_rxpkts)
+{
+	size_t i = 0, j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	while (i < num_txpkts) {
+		if (read8(pd->base + SSPSR) & SSPSR_TNF)
+			/* tx 1 packet */
+			write16(wdat[i++], pd->base + SSPDR);
+	}
+
+	do {
+		while ((read8(pd->base + SSPSR) & SSPSR_RNE)
+			&& (j < *num_rxpkts))
+			/* rx 1 packet */
+			rdat[j++] = read16(pd->base + SSPDR);
+	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
```

Comment:
That was my initial thought too, that read/write at the same time is better, but as @jenswi-linaro had pointed out, it made the code a bit more complicated than necessary. Performance-wise, it's also not necessarily better because previously we had write + read + is_fifo_full checking in one loop and then read in another loop, whereas now we just have a simple write and then read loop, with the fifo checking naturally integrated into the write loop. The new approach is also more conventional and inline with other pl022/spi drivers out there in the open.